src/main/scala/org/broadinstitute/hail/methods/ImputeSexPlink.scala

```diff
@@ -1,129 +1,77 @@
-          require(v == v2)
-          (v, va, gs, vinfo)
-        }
-      }
-      .filter { case (v, va, gs, (maf, nCalled)) =>
-        maf.isDefined && maf.get >= localMafThreshold && nCalled > 0
+  def apply(vds: VariantDataset,
+    mafThreshold: Double,
+    excludePar: Boolean,
+    fMaleThreshold: Double,
+    fFemaleThreshold: Double,
+    popFrequencyExpr: Option[String]): Map[String, Annotation] = {
+
+    val query = popFrequencyExpr.map { code =>
+      val (t, f) = vds.queryVA(code)
+      t match {
+        case TDouble => f
+        case other => fatal(s"invalid population frequency.  Expected Double, but got `$other'")
       }
-      .map { case (v, va, gs, (maf, nCalled)) => (v, va, gs)})
-  }
-
-  private def inbreedingCoefficients2: RDD[(String, InbreedingCombiner)] = {
-    filteredXchrRDD2.aggregateBySampleWithVariantInfo(new InbreedingCombiner)(
-      (comb, v, va, s, sa, g, a) => comb.addCount(g, a.asInstanceOf[Double]),
-      (comb1, comb2) => comb1.combineCounts(comb2),
-      populationParameters.map{case (v, (maf, nCalled)) => (v, maf)})
-  }
-
-  private def inbreedingCoefficients: RDD[(String, InbreedingCombiner)] = {
-    val sparkContext = xChrVds.sparkContext
-    val sampleIdsBC = xChrVds.sampleIds
-
-    val serializer = SparkEnv.get.serializer.newInstance()
-    val zeroBuffer = serializer.serialize(new InbreedingCombiner)
-    val zeroArray = new Array[Byte](zeroBuffer.limit)
-    zeroBuffer.get(zeroArray)
-
-    filteredXchrRDD
-      .mapPartitions { (it: Iterator[(Variant, Annotation, Iterable[Genotype], Any)]) =>
-        val serializer = SparkEnv.get.serializer.newInstance()
-        def copyZeroValue() = serializer.deserialize[InbreedingCombiner](ByteBuffer.wrap(zeroArray))
-        val arrayZeroValue = Array.fill[InbreedingCombiner](sampleIdsBC.length)(copyZeroValue())
-
-        sampleIdsBC.iterator
-          .zip(it.foldLeft(arrayZeroValue) { case (acc, (v, va, gs, (maf: Double, nCalled: Int))) =>
-            for ((g, i) <- gs.iterator.zipWithIndex)
-              acc(i) = acc(i).addCount(g, maf)
-            acc
-          }.iterator)
-      }.foldByKey(new InbreedingCombiner)((ibc1, ibc2) => ibc1.combineCounts(ibc2))
-
-    .map { case (s, ibc) => (s,ibc)}
-  }
+    }
 
-  def result(fFemaleThreshold: Double, fMaleThreshold: Double) = {
-    inbreedingCoefficients.map { case (s, ibc) => (s, Annotation(ibc.F, ibc.E, ibc.O, ibc.N, ibc.T, determineSex(ibc, fFemaleThreshold, fMaleThreshold))) }
+    vds.filterVariants((v: Variant, va: Annotation, gs: Iterable[Genotype]) => (v.contig == "X" || v.contig == "23") && !(excludePar && v.inParX))
+      .mapAnnotations((v: Variant, va: Annotation, gs: Iterable[Genotype]) =>
+        query.map(_.apply(va).orNull)
+          .getOrElse({
+            var nAlt = 0
+            var nTot = 0
+            for (g <- gs) {
+              g.gt.foreach { call =>
+                nAlt += call
+                nTot += 2
+              }
```

Comment:
Does this work in general? I thought g.gt could be anything in the range of 0 to nGenotypes.